File: 2023/java/D02P2.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.Arrays;
import java.util.stream.Stream;

public class d02p2 {
    private static final int ZERO = 48;

    public static void main(String[] args) throws IOException {
        System.out.println("Day 2 - Part 2");

        BufferedReader reader = new BufferedReader(new FileReader(new File("../_input/day02.txt")));
        
        int total = 0;

        String line;
        while((line = reader.readLine()) != null) {
            total += process(line.trim());
        }

        System.out.println("Total: " + total);
    }

    private static int process(String str) {
        Game game = getGame(str);

        int maxRed = 0;
        int maxGreen = 0;
        int maxBlue = 0;

        for(var s : game.subsets) {
            maxRed = Math.max(maxRed, s.red);
            maxGreen = Math.max(maxGreen, s.green);
            maxBlue = Math.max(maxBlue, s.blue);
        }

        return maxRed * maxGreen * maxBlue;
    }

    private static Game getGame(String str) {
        int colon = str.indexOf(":");
        String rawSubSets = str.substring(colon + 2);
        Stream<String> ss = Arrays.stream(rawSubSets.split(";")).map(i -> i.trim());
        SubSet[] sets = ss.map(i -> getSubSet(i)).toArray(size -> new SubSet[size]);

        return new Game(sets);
    }

    private static SubSet getSubSet(String str) {
        String[] sets = Arrays.stream(str.split(",")).map(i -> i.trim()).toArray(size -> new String[size]);

        int red = 0;
        int green = 0;
        int blue = 0;

        for(var s : sets) {
            if(s.endsWith("red"))
                red += stringToInt(s.substring(0, s.length() - 4));
            else if(s.endsWith("green"))
                green += stringToInt(s.substring(0, s.length() - 6));
            else if(s.endsWith("blue"))
                blue += stringToInt(s.substring(0, s.length() - 5));  
            else {
                throw new RuntimeException("Unknown color");
            }
        }

        return new SubSet(red, green, blue);
    }

    private static int stringToInt(String str) {
        int acc = 0;
        for(int i = 0; i < str.length(); i++) {
            int c = str.charAt(i);
            acc += Math.pow(10.0, str.length() - i - 1) * (c - ZERO);
        }
        return acc;
    }

    private record Game(SubSet[] subsets) {}
    private record SubSet(int red, int green, int blue) {}
}

File: 2023/zig/d02p2.zig
const std = @import("std");
const allocator = std.heap.page_allocator;

const ZERO: u8 = 48;

const Game = struct { subsets: std.ArrayList(SubSet) };
const SubSet = struct { red: u32, green: u32, blue: u32 };

pub fn main() anyerror!void {
    std.debug.print("Day 2 - Part 2\n", .{});

    const file = try std.fs.cwd().openFile("../_input/day02.txt", .{});
    defer file.close();

    var buf_reader = std.io.bufferedReader(file.reader());
    const in = buf_reader.reader();

    var total: u32 = 0;

    var buffer: [1024]u8 = undefined;
    while (try in.readUntilDelimiterOrEof(&buffer, '\n')) |line| {
        total += process(line);
    }

    std.debug.print("Total: {}\n", .{total});
}

fn process(str: []const u8) u32 {
    var game = get_game(str);

    var max_red: u32 = 0;
    var max_green: u32 = 0;
    var max_blue: u32 = 0;

    const ss = unwrap([]SubSet, game.subsets.toOwnedSlice());
    for (ss) |sub| {
        max_red = @max(max_red, sub.red);
        max_green = @max(max_green, sub.green);
        max_blue = @max(max_blue, sub.blue);
    }

    return max_red * max_green * max_blue;
}

fn get_game(str: []const u8) Game {
    const colon = std.mem.indexOf(u8, str, ":").?;
    const raw_subsets = str[(colon + 1)..str.len];

    var ss = std.mem.splitAny(u8, raw_subsets, ";");
    var subsets = std.ArrayList(SubSet).init(allocator);
    while (ss.next()) |s| {
        const set = get_subset(s);
        unwrap(void, subsets.append(set));
    }

    return Game{ .subsets = subsets };
}

fn get_subset(str: []const u8) SubSet {
    var subs = std.mem.splitAny(u8, str, ",");

    var red: u32 = 0;
    var green: u32 = 0;
    var blue: u32 = 0;

    while (subs.next()) |sset| {
        const s = std.mem.trim(u8, sset, " \t\n");
        if (std.mem.endsWith(u8, s, "red")) {
            red += str_to_u32(s[0..(s.len - 4)]);
        } else if (std.mem.endsWith(u8, s, "green")) {
            green += str_to_u32(s[0..(s.len - 6)]);
        } else if (std.mem.endsWith(u8, s, "blue")) {
            blue += str_to_u32(s[0..(s.len - 5)]);
        } else {
            std.debug.panic("Unknown color", .{});
        }
    }

    return SubSet{ .red = red, .green = green, .blue = blue };
}

fn str_to_u32(str: []const u8) u32 {
    var acc: u32 = 0;
    for (str, 0..) |c, i| {
        acc += (unwrap(u32, std.math.powi(u32, 10, @as(u32, @intCast(str.len - i - 1))))) * (c - ZERO);
    }
    return acc;
}

fn unwrap(comptime T: type, val: anyerror!T) T {
    if (val) |value| {
        return value;
    } else |err| {
        std.debug.panic("Panicked at Error: {any}", .{err});
    }
}
